File: linked_list/92.reverse-linked-list-ii.cpp
/*
 * @lc app=leetcode id=92 lang=cpp
 *
 * [92] Reverse Linked List II
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
 public:
  ListNode *reverseBetween(ListNode *head, int left, int right) {
    ListNode dummy, *curr = &dummy;
    curr->next = head;

    int i = 1;
    while (i++ < left and curr) {
      curr = curr->next;
    }
    ListNode *prev_left = curr;

    i = left - 1;
    while (i++ <= right and curr) {
      curr = curr->next;
    }
    ListNode *after_right = curr;

    prev_left->next = helper(prev_left->next, after_right);

    return dummy.next;
  }

  ListNode *helper(ListNode *start, ListNode *end) {
    ListNode *prev = end;
    while (start != end) {
      ListNode *next = start->next;
      start->next = prev;
      prev = start;
      start = next;
    }

    return prev;
  }
};
// @lc code=end

File: 92.reverse-linked-list-ii.py
#
# @lc app=leetcode id=92 lang=python3
#
# [92] Reverse Linked List II
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
from typing import Optional, Tuple


class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        if not head:
            return head
        dummy = ListNode(-1, head)
        prev = dummy
        count = 1
        
        while prev and count < left:
            count += 1
            prev = prev.next
        
        prev.next = self.doReverse(prev.next, right - count + 1)
        return dummy.next
        
        
    def doReverse(self, start: ListNode, count: int) -> ListNode:
        curr = start
        prev = None
        while curr and count:
            count -= 1
            next = curr.next
            curr.next = prev
            prev = curr
            curr = next
        start.next = curr
        
        return prev


        
    
# @lc code=end

